dask/array/optimization.py

```diff
@@ -290,3 +299,182 @@ def fuse_slice(a, b):
+                    top_layers.add(dep)
+                    for d in sharedict.dependencies.get(dep, ()):
+                        if len(dependents[d]) <= 1:
+                            deps.add(d)
+                        else:
+                            stack.append(d)
+                else:
+                    stack.append(dep)
+            new_layer = rewrite_atop([layers[l] for l in top_layers])
+            out[layer] = new_layer
+            dependencies[layer] = {k for k, v in new_layer.indices if v is not None}
+        else:
+            out[layer] = layers[layer]
+            dependencies[layer] = sharedict.dependencies.get(layer, set())
+            stack.extend(sharedict.dependencies.get(layer, ()))
+
+    return ShareDict(out, dependencies)
+
+
+def rewrite_atop(inputs):
+    """ Rewrite a stack of atop expressions into a single atop expression
+
+    Given a set of TOP layers, combine them into a single layer.  The provided
+    layers are expected to fit well together.  That job is handled by
+    ``optimize_atop``
+
+    Parameters
+    ----------
+    inputs : List[TOP]
+
+    Returns
+    -------
+    top : TOP
+
+    See Also
+    --------
+    optimize_atop
+    """
+    from dask.core import reverse_dict
+    from dask.array.core import TOP
+    inputs = {inp.output: inp for inp in inputs}
+    dependencies = {inp.output: {d for d, v in inp.indices if v is not None and d in inputs}
+                    for inp in inputs.values()}
+    dependents = reverse_dict(dependencies)
+
+    new_index_iter = iter('ABCDEFGHIJKLMNOPQRSTUVWXYZ')
+
+    [root] = [k for k, v in dependents.items() if not v]
+
+    indices = list(inputs[root].indices)
+    new_axes = inputs[root].new_axes
+    concatenate = inputs[root].concatenate
+
+    dsk = dict(inputs[root].dsk)
+
+    seen = set()
+
+    changed = True
+    while changed:
+        changed = False
+        for i, (dep, ind) in enumerate(indices):
+            if ind is None:
+                continue
+            if dep not in inputs:
+                continue
+
+            changed = True
+
+            # Replace _n with dep name in existing tasks
+            # (inc, _0) -> (inc, 'b')
+            dsk = {k: subs(v, {atop_token(i): dep}) for k, v in dsk.items()}
+
+            # Remove current input from input indices
+            # [('a', 'i'), ('b', 'i')] -> [('a', 'i')]
+            _, current_dep_indices = indices.pop(i)
+            sub = {atop_token(i): atop_token(i - 1) for i in range(i + 1, len(indices) + 1)}
+            dsk = subs(dsk, sub)
+
+            # Change new input_indices to match give index from current computation
+            # [('c', j')] -> [('c', 'i')]
+            new_indices = inputs[dep].indices
+            sub = dict(zip(inputs[dep].output_indices, current_dep_indices))
+            contracted = {x for _, j in new_indices if j is not None for x in j if x not in inputs[dep].output_indices}
```

Comment:
Would you mind wrapping long comprehensions like this? I find long comprehensions more readable if they break at expression changes. This one I'd wrap as:

```
contracted = {x for _, j in new_indices
              if j is not None
              for x in j
              if x not in inputs[dep].output_indices}
```

Reply:
Done in a few places.  Agreed that it's more readable.  I'll keep an eye out for others.